### PR TITLE
wallet: feature: transfer amount with fee included

### PR DESCRIPTION
**Pull request by jeffro256**
To transfer ~5 XMR to an address such that your balance drops by exactly 5 XMR, provide a `subtractfeefrom` flag to the `transfer` command. For example:

    transfer 76bDHojqFYiFCCYYtzTveJ8oFtmpNp3X1TgV2oKP7rHmZyFK1RvyE4r8vsJzf7SyNohMnbKT9wbcD3XUTgsZLX8LU5JBCfm 5 subtractfeefrom=all

If my walet balance was exactly 30 XMR before this transaction, it will be exactly 25 XMR afterwards and the destination address will receive slightly less than 5 XMR. You can manually select which destinations fund the transaction fee and which ones do not by providing the destination index. For example:

    transfer 75sr8AAr... 3 74M7W4eg... 4 7AbWqDZ6... 5 subtractfeefrom=0,2

This will drop your balance by exactly 12 XMR including fees and will spread the fee cost equally over destinations with addresses `75sr8AAr...` and `7AbWqDZ6...`.

This feature was requested by @LocalMonero.

### CODE REVIEW

**Comment by LocalMonero** [comment]
This closes #7980

**Comment by jeffro256** [comment]
Depends on #8888

**Comment by woodser** [comment]
This is a needed feature for application developers which I will be using too.

Is anything else needed before merging? @luigi1111

**Comment by jeffro256** [comment]
@woodser if you would like to see this happen, would you mind reviewing #8888 as well if you’d like?

**Comment by selsta** [comment]
@woodser it needs an approval for this and #8888

**Comment by woodser** [comment]
@jeffro256 Please rebase on master.

**Review by woodser — CHANGES_REQUESTED** [review]

File: src/wallet/wallet_rpc_server_commands_defs.h
Comment (woodser) [reply src/wallet/wallet_rpc_server_commands_defs.h:580]: This field will not support the `all` option over the API.

That's fine with me to avoid the complexity of using a string to store "all" or CSV.

Just checking that this is the preferred approach, that the client will pass all destination indices manually, and `all` will not be supported.

File: src/wallet/wallet_rpc_server.cpp
Comment (jeffro256) [reply src/wallet/wallet_rpc_server.cpp:1095]: For me, it doesn't make as much sense to add this to the split transfer interface since the fees have to be paid out of certain destinations. If those certain destinations are split over multiple transactions, it's unclear which destinations should pay for transactions if a transaction doesn't contain any "subtractable" destinations. Also, this feature can be expanded upon in a future PR for reviewer's sake, and I'd imagine this feature is mainly useful for small transfers (like customers paying for an item with fee included).

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp]: fixed, thanks

File: src/simplewallet/simplewallet.cpp
Comment (jeffro256) [reply src/simplewallet/simplewallet.cpp]: fixed!

File: src/wallet/wallet_rpc_server_commands_defs.h
Comment (jeffro256) [reply src/wallet/wallet_rpc_server_commands_defs.h:580]: Yes this is the preferred option for me since it simplifies the RPC interface. If you want to do "all", just list out all the destination indexes which is pretty trivial IMO

Comment (jeffro256) [reply src/wallet/wallet_rpc_server_commands_defs.h:580]: >  This needs added to COMMAND_RPC_TRANSFER_SPLIT too.

As for this, see https://github.com/monero-project/monero/pull/8861#discussion_r1257514658

**Comment by jeffro256** [comment]
> @jeffro256 Please rebase on master.

There was an issue but rebasing was fixed

**Comment by woodser** [comment]
@jeffro256 Do you mind to PR this on the release branch in addition to the master branch, so I can cherry-pick only what is necessary for my next release?

Otherwise I would have to base on master which has diverged from the release branch, so it's unofficial and could be risky.

File: src/wallet/wallet_rpc_server.cpp
Comment (woodser) [reply src/wallet/wallet_rpc_server.cpp:1095]: simplewallet.cpp already permits applying this feature across split transactions, so either that behavior needs to be well defined and supported or completely disabled.

Then monero-wallet-rpc should be consistent. So if we don't support `transfer_split` in monero-wallet-rpc now, this feature should be disabled for split transactions in simplewallet too.

As far as I can tell, the expected behavior with split transactions should be clear, where the fee is split across any specified destinations in each transaction, and the entire call fails if any transaction cannot be created with subtractable destinations, so my preference is to support this.

**Comment by jeffro256** [comment]
In the latest commit, I explicitly disabled using this feature split over multiple transactions to avoid user confusion.

**Review by vtnerd — CHANGES_REQUESTED** [review]

File: src/simplewallet/simplewallet.cpp
Comment (selsta) [reply src/simplewallet/simplewallet.cpp:524]: https://github.com/monero-project/monero/pull/8922

This PR is still open, so if you changed your opinion I'd appreciate a comment.

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp:9790]: This might be a better design...

**Comment by woodser** [comment]
@jeffro256 mind updating the [release PR](https://github.com/monero-project/monero/pull/8945) too?

**Comment by woodser** [comment]
This pull request provides critical functionality to deduct and split the mining fee among transfer amounts. Haveno relies heavily on this feature in its trade protocol. Happy to help move this forward however possible.

File: src/simplewallet/simplewallet.cpp
Comment (nahuhh) [reply src/simplewallet/simplewallet.cpp:524]: 8922 is merged now @vtnerd @jeffro256

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp:9898]: I don't remember... modified

**Comment by jeffro256** [comment]
Rebased. Will acknowledge @vtnerd's comments soon...

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp:9958]: This can't underflow because it's guarded by the condition `d.amount` in the `if` branch

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp:9947]: You're right, I had to re-write this section.

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp:9894]: Okay after looking at it, I think it's a better option to not do this since we would need to update all the code which uses `tx_destination_entry` to construct transactions, including the multisig ringct tx builder class.

File: src/wallet/wallet2.cpp
Comment (vtnerd) [inline src/wallet/wallet2.cpp:10557]: This is an overestimate? Will this cause spurious failures (I'm not familiar with `wallet2::sanity_check`?

**Review by vtnerd — APPROVED** [review]
One minor question.

File: src/wallet/wallet2.cpp
Comment (jeffro256) [reply src/wallet/wallet2.cpp:10557]: I think it's just wrong assuming that the fees for the different `pending_tx` can be different.

Comment (jeffro256) [reply src/wallet/wallet2.cpp:10557]: Wait, no that's fine since we decided we wouldn't support this feature for split transactions. So this shouldn't be an overestimate. If anything it's an underestimate since the minimum should be `target - fee / num_subtractable`, right?

Comment (jeffro256) [reply src/wallet/wallet2.cpp]: Made the sanity check accurate to within one piconero for subtractable destinations

**Comment by woodser** [comment]
Thanks for your updates @jeffro256.

I can test it in the libraries and applications once the [release PR](https://github.com/monero-project/monero/pull/8945) is rebased.

**Comment by jeffro256** [comment]
@woodser I won't be able to rebase onto the release branch until c++17 is added to the release branch. Relevant PR: #8922

**Comment by jeffro256** [comment]
I'll keep it in draft mode as long as this PR is awaiting approval

**Review by selsta — COMMENTED** [review]
Please bump WALLET_RPC_VERSION_MINOR

**Comment by jeffro256** [comment]
Done

**Review by selsta — APPROVED** [review]

**Comment by jeffro256** [comment]
Oh shoot you're right. I always get those mixed up

**Review by selsta — APPROVED** [review]

**Pull request merged**
